### PR TITLE
Create defcustom for allowing free movement of point in task buffers

### PR DESCRIPTION
**Pull request by gempesaw**

### CODE REVIEW

File: grunt.el [hunk 1]
```diff
@@ -126,6 +126,21 @@ argument when invoking `grunt-exec'."
   :type 'boolean
   :group 'grunt)
 
+(defcustom grunt-move-task-point nil
```
Comment (domtronn) [inline grunt.el]: I'd maybe call this something like `grunt-task-auto-scroll`?

That feels a bit more descriptive of the variables purpose :) 
What do you think?

**Comment by domtronn** [comment]
Hey @gempesaw! This looks good, but it'd be nice if we can avoid the duplication of the process filter function like you mentioned in another comment. I've had a bit of a play around, and _you were right!_ looks like it is time for a macro!!

So rather than toggling on the `set-process-filter` call, you could try something like this
```cl
(defun grunt--apply-ansi-color (proc string)
  "Filter to function for process PROC to apply ansi color to STRING."
  (when (buffer-live-p (process-buffer proc))
    (with-current-buffer (process-buffer proc)
      (let ((inhibit-read-only t))
        (grunt--filter-body
         grunt-move-task-point
         ;; Insert the text, advancing the process marker.
         (goto-char (process-mark proc))
         (insert string)
         (ansi-color-apply-on-region (process-mark proc) (point))
         (set-marker (process-mark proc) (point)))))))

(defmacro grunt--filter-body (scroll &rest body)
  "Switch on SCROLL through the output of a process while executing BODY."
  `(if ,scroll (progn ,@body) (save-excursion ,@body)))
```
_(not sure about the function name... I just wanted something that worked)_

I think it's worth noting here that using the `grunt-move-task-point` variable inside the macro doesn't work, I think because it's only evaluated once? So if the variable changes, the macro does not... But passing it through as `scroll` causes it to be reevaluated on each call. :)

And then when it comes to testing...... I've been using `ert-async` on my [`button-links-done-proper`](https://github.com/gempesaw/grunt.el/tree/button-links-done-proper) branch, and it allows you to run `grunt-exec` asynchronously, and do your assertions when it's completed! Which means the filter functions are actually called :P

Here's another snippet
```cl
(ert-deftest-async should-set-save-excursion-process-filter (done)
  (with-grunt-sandbox
   (let ((call-with-scroll 5)
         (grunt-move-task-point t))
     (noflet ((completing-read (&rest any) "build")
              (grunt--filter-body (scroll &rest any) (setq call-with-scroll scroll)))
       (let ((proc (grunt-exec)))
         (set-process-sentinel proc
           `(lambda (&rest any) (should ,call-with-scroll) (funcall done))))))))
```
So, there's a bit of setup up to be done for this test though... You'd need to add `ert-async` to the `CASK` file, require `ert-async` in `grunt-test.el`, and I think that's it? I set `call-with-scroll` to 5, so you can tell if you get false positives with the `nil` `t` checks etc...

File: grunt.el (hunk 1 above)
Comment (gempesaw) [inline grunt.el]: oh yea that is way better. i couldn't think of anything good. having changed the name, the meanings would also change, i presume:

defaults to `t` for keeping point at (point-max)
nil means let the user decide what to do with the point

**Comment by gempesaw** [comment]
> I think it's worth noting here that using the grunt-move-task-point variable inside the macro doesn't work, I think because it's only evaluated once? So if the variable changes, the macro does not... But passing it through as scroll causes it to be reevaluated on each call. :) 

ahh that's tricky, much obliged that you caught that! I was curious why the macro signature asked for the scroll arg, but that makes sense

> Here's another snippet

cool, I will try that out. I was curious about how to test that part and basically just punted, but I'll give it a shot :D

**Comment by gempesaw** [comment]
blergh, spent a while fiddling with this but couldn't get the movement to track the process-mark iff the point was already at the previous process-mark.

on the plus side i was reading the process-filters from comint-mode and compilation-mode and they're way more complicated than what we've got going on. i am thinking vaguely that it should be using `(define-compilation-mode ...`, since that's basically what we want: a compilation buffer will handle the output filters for us, the colorization will pick up the same colorization setup that your compilation buffers already use, and the error handling that compilation buffers get would get inherited automatically. also, they decided to name their variable `compilation-scroll-output` which is ... pretty close to what you came up with :D. but with better docs, more options, and way more thought out than what I was trying to kludge together here. :)

well, in the meantime, i'll give up on this and merge that other PR.

**Comment by domtronn** [comment]
Sure, I'm having trouble with the tests now actually.. I might have a look and see if I can get this working though :)

I had a go at inheriting from compilation mode, but to be honest, it didn't really seem to add much... It didn't match any of the error paths, it didn't apply colours to the output _(at least not the ansi colors that you get when grunt prints to the console)_ and the `compilation-scroll-output` did nothing :P 

I might be doing it wrong! And if we can get those benefits for free then brill :) but I might try and at least emulate its behaviour as much as possible in our stuff

**Pull request closed without merging**